asf_hp_installer_chatbot/pipeline/heat_pump_chatbot_post_embedding.py

```diff
@@ -0,0 +1,302 @@
+    The API key is required for authentication with OpenAI's services. The model name determines which version
+    of GPT (Generative Pre-trained Transformer) will be used for generating responses.
+    The temperature parameter controls the randomness of the output, with lower values producing more deterministic
+    and predictable text, and higher values resulting in more varied and creative responses.
+
+
+    Returns:
+        ChatOpenAI: A ChatOpenAI object with the specified API key, model name, and temperature.
+    """
+    return ChatOpenAI(openai_api_key=openai.api_key, model_name=model, temperature=temp)
+
+
+def get_retrieval_qa(llm: ChatOpenAI, vectorstore: Pinecone) -> RetrievalQA:
+    """
+    Returns a RetrievalQA object for the specified language model and vector store.
+    When "Retrieval QA" uses the chain_type "stuff" in LangChain, it indicates a process where the system searches through external
+    documents to find relevant information. Once this information (the "stuff") is identified, it is “stuffed” (fits within context window)
+    into the LLM.
+    Args:
+        llm (ChatOpenAI): The language model to use.
+        vectorstore (Pinecone): The vector store to use.
+
+    Returns:
+        RetrievalQA: A RetrievalQA object for the specified language model and vector store.
+    """
+    return RetrievalQA.from_chain_type(
+        llm=llm, chain_type="stuff", retriever=vectorstore.as_retriever()
+    )
+
+
+def get_sender_hash(sender_number: str) -> str:
+    """
+    Generates and returns a SHA256 hash of the sender's number. This is done so as to anonymise the sender's number.
+
+    Args:
+        sender_number (str): The sender's number.
+
+    Returns:
+        str: A SHA256 hash of the sender's number.
+    """
+    sender_number = sender_number.replace("whatsapp:", "")
+    return hashlib.sha256(sender_number.encode()).hexdigest()
+
+
+def get_timestamp() -> str:
+    """
+    Returns the current timestamp in the format 'YYYY-MM-DD HH:MM:SS'. Useful to have a timestamp to understand ordering of messages.
+
+    Returns:
+        str: The current timestamp.
+    """
+    return datetime.now().strftime("%Y-%m-%d %H:%M:%S")
+
+
+def append_to_csv(
+    filename: str, sender_hash: str, incoming_msg: str, answer: str, timestamp: str
+):
+    """
+    Appends a row of data to a CSV file.
+
+    Args:
+        filename (str): The name of the CSV file.
+        sender_hash (str): The SHA256 hash of the sender's number.
+        incoming_msg (str): The incoming message.
+        answer (str): The chatbot's response.
+        timestamp (str): The timestamp of the message.
+    """
+    file_exists = os.path.isfile(filename)
+    with open(filename, "a", newline="") as f:
+        writer = csv.writer(f)
+        if not file_exists:
+            writer.writerow(["ID", "Question", "Answer", "Timestamp"])  # header
+        writer.writerow([sender_hash, incoming_msg, answer, timestamp])
```

Comment:
Should this be:
```suggestion
    if not file_exists:
        writer = csv.writer(f)
        with open(filename, "w", newline="") as f:
            writer.header(["ID", "Question", "Answer", "Timestamp"])  # header
    with open(filename, "a", newline="") as f:
        writer = csv.writer(f)
        writer.writerow([sender_hash, incoming_msg, answer, timestamp])
```
when i ran the code originally i got an error that this outputs file did not exist so if it doesnt it needs to be created first.

Reply:
The function append_to_csv uses the open function with mode "a" (append), which should create the file if it doesn't exist. However, if the directory where the file is supposed to be created doesn't exist, an error will occur? Perhaps that was the issue?